src/listener.js

```diff
@@ -1,156 +1,242 @@
+          throw new AllListenersFailedError()
+        }
+      })
   }
 
-  let ipfsId
-  let listeningAddr
-
-  listener.listen = (ma, callback) => {
-    listeningAddr = ma
-    if (includes(ma.protoNames(), 'ipfs')) {
-      ipfsId = getIpfsId(ma)
-      listeningAddr = ma.decapsulate('ipfs')
+  /**
+   * Removes the server from tracking and performs cleanup.
+   * If all servers have been closed, `close` will be emitted by
+   * the listener.
+   * @private
+   * @param {net.Server} server
+   */
+  _removeServer (server) {
+    // only emit if we're not listening
+    if (!this._isListening()) {
+      this.emit('close')
     }
-
-    const lOpts = listeningAddr.toOptions()
-    log('Listening on %s %s', lOpts.port, lOpts.host)
-    return server.listen(lOpts.port, lOpts.host, callback)
+    this._servers.delete(server)
+    server.removeAllListeners()
   }
 
-  listener.getAddrs = (callback) => {
+  /**
+   * Return the addresses we are listening on
+   * @throws
+   * @returns {Array<Multiaddr>}
+   */
+  getAddrs () {
     const multiaddrs = []
-    const address = server.address()
-
-    if (!address) {
-      return callback(new Error('Listener is not ready yet'))
-    }
-
-    // Because TCP will only return the IPv6 version
-    // we need to capture from the passed multiaddr
-    if (listeningAddr.toString().indexOf('ip4') !== -1) {
-      let m = listeningAddr.decapsulate('tcp')
-      m = m.encapsulate('/tcp/' + address.port)
-      if (ipfsId) {
-        m = m.encapsulate('/ipfs/' + ipfsId)
-      }
+    this._servers.forEach(server => {
+      const address = server.address()
 
-      if (m.toString().indexOf('0.0.0.0') !== -1) {
+      if (address.address === '0.0.0.0') {
         const netInterfaces = os.networkInterfaces()
         Object.keys(netInterfaces).forEach((niKey) => {
           netInterfaces[niKey].forEach((ni) => {
-            if (ni.family === 'IPv4') {
-              multiaddrs.push(multiaddr(m.toString().replace('0.0.0.0', ni.address)))
+            if (address.family === ni.family) {
+              multiaddrs.push(
+                multiaddr.fromNodeAddress({
+                  ...address,
+                  address: ni.address
+                }, 'tcp')
+              )
             }
           })
         })
       } else {
-        multiaddrs.push(m)
+        multiaddrs.push(multiaddr.fromNodeAddress(address, 'tcp'))
       }
+    })
+
+    if (multiaddrs.length === 0) {
+      throw new Error('Listener is not ready yet')
     }
 
-    if (address.family === 'IPv6') {
-      let ma = multiaddr('/ip6/' + address.address + '/tcp/' + address.port)
-      if (ipfsId) {
-        ma = ma.encapsulate('/ipfs/' + ipfsId)
-      }
+    return multiaddrs
+  }
 
-      multiaddrs.push(ma)
+  /**
+   * Handler for new sockets from `net.createServer`
+   * @param {net.Socket} socket
+   */
+  _onSocket (socket) {
+    // Avoid uncaught errors caused by unstable connections
+    socket.on('error', noop)
```

Comment:
Log the error at least?

Reply:
This was in the original code, do you think we should really just log/ignore the error, or should we throw it?

Reply:
@jacobheun do you have a strong feeling about this one way or the other (log vs throw)?

Reply:
I would log it. Ultimately, that should only trigger if the socket errors before we've started processing it, so the application should never get notified of the incoming connection. Throwing  an error doesn't make much sense since the application doesn't need to take any action.

Reply:
Right that's true 👍